### PR TITLE
Add health check routines to Stanza

### PR DESCRIPTION
**Pull request by rjowcq**
Add improved "health check" tests based on "BATIS: Bootstrapping, Autonomous Testing, and Initialization System for Quantum Dot Devices".  We add a noise floor measurement and an enhanced leakage test that sweeps in both the positive and negative directions from the initial voltages to detect asymmetric leakage behavior.

• New analysis module: Added curve fitting capabilities for pinchoff characterization including the fit_pinchoff_parameters()
  function, quality assessment criteria (R², NRMSE), and data preprocessing utilities

• Health check routines: Implemented 5 device health check routines - noise floor measurement, leakage
testing, global accumulation, reservoir characterization, and finger gate characterization - adapted from BATIS quantum dot testing methodology

• Routine framework: Added core infrastructure for defining and executing measurement routines with decorators, context objects, and a runner system for organizing measurement workflows

• QDAC2 driver enhancements: Added batch measurement support allowing multiple channels to be measured in a single query for improved performance

• Device sweep API changes: Modified sweep methods to return voltage measurements as list-of-lists (one list per sweep point) instead of flattened lists, enabling proper multi-gate sweep tracking

### CODE REVIEW

File: stanza/analysis/fitting.py [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+DEFAULT_C_MARGIN = 10.0
+DEFAULT_BOUNDS = [(1e-8, 1.0), (-20.0, 20.0), (-10.0, 10.0)]
+
+
+@dataclass
+class PinchoffFitResult:
+    """Result of pinchoff curve fitting.
+
+    Attributes:
+        vp: Pinchoff voltage (where device turns off)
```
Comment (JoelPendleton) [inline stanza/analysis/fitting.py]: more verbose names wouldn't hurt here, e.g. v_pinch_off, v_transition, v_cut_off

File: stanza/drivers/qdac2.py [hunk 2]
```diff
@@ -185,13 +193,7 @@ def _initialize_channels(self, channel_configs: dict[str, ChannelConfig]) -> Non
...
     def prepare_measurement(self) -> None:
         """Prepare the measurement."""
-        # Set current measurement range
-        channels_str = ",".join(str(ch) for _, ch in self.measurement_channels)
-        self.driver.write(
-            f"sens:rang {str(self.current_range).lower()},(@{channels_str})"
-        )
-        # Set the integration time
-        self.driver.write(f"sens:aper {self.measurement_aperature_s},(@{channels_str})")
+        pass
```
Comment (JoelPendleton) [inline stanza/drivers/qdac2.py]: why does this exist, if it does nothing?

File: stanza/analysis/fitting.py [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+
+    HAS_SCIPY = True
+except ImportError:
+    HAS_SCIPY = False
+
+from stanza.analysis.preprocessing import normalize
+
+# Constants for parameter bounds computation
+DEFAULT_C_MARGIN = 10.0
+DEFAULT_BOUNDS = [(1e-8, 1.0), (-20.0, 20.0), (-10.0, 10.0)]
```
Comment (JoelPendleton) [inline stanza/analysis/fitting.py:17]: Do these assume the data is normalized?

File: stanza/drivers/qdac2.py [hunk 4]
```diff
@@ -152,6 +152,14 @@ def __init__(
         super().__init__(instrument_config)
         self._initialize_channels(channel_configs)
 
+        # Set current measurement range
+        channels_str = ",".join(str(ch) for _, ch in self.measurement_channels)
+        self.driver.write(
+            f"sens:rang {str(self.current_range).lower()},(@{channels_str})"
+        )
+        # Set the integration time
+        self.driver.write(f"sens:aper {self.measurement_aperature_s},(@{channels_str})")
```
Comment (JoelPendleton) [inline stanza/drivers/qdac2.py]: Can we use nplc too?

File: stanza/routines/builtins/health_check.py [hunk 5]
```diff
@@ -0,0 +1,657 @@
+"""Device health check routines for quantum dot devices.
+
+Health check tests adapted from:
```
Comment (JoelPendleton) [inline stanza/routines/builtins/health_check.py]: I'd say 'improvement of', there is a lot added.

File: stanza/routines/builtins/health_check.py [hunk 6]
```diff
@@ -0,0 +1,657 @@
...
+from stanza.models import GateType
+from stanza.routines import RoutineContext, routine
+
+logger = logging.getLogger(__name__)
+
+DEFAULT_SETTLING_TIME_S = 10
+
+
+@routine
+def noise_floor_measurement(
```
Comment (JoelPendleton) [inline stanza/routines/builtins/health_check.py:28]: All voltages should be zero on all electrodes

File: stanza/routines/builtins/health_check.py [hunk 7]
```diff
@@ -0,0 +1,657 @@
...
+
+    Notes:
+        - Device is always returned to initial voltages in the finally block
+        - Leakage matrix is symmetric, only upper triangle is checked
+        - Both voltage bounds are tested regardless of failures
+        - Current differences below min_current_threshold are skipped to avoid noise
+        - Voltage bounds are determined from device channel configurations
+    """
+
+    leakage_threshold_resistance = int(leakage_threshold_resistance)
```
Comment (JoelPendleton) [inline stanza/routines/builtins/health_check.py:128]: You don't need to make it an int. It should already be an int

File: stanza/routines/builtins/health_check.py [hunk 8]
```diff
@@ -0,0 +1,657 @@
...
+    leakage_threshold_count = int(leakage_threshold_count)
```
Comment (JoelPendleton) [inline stanza/routines/builtins/health_check.py:129]: same as above

File: stanza/routines/builtins/health_check.py [hunk 9]
```diff
@@ -0,0 +1,657 @@
...
+    num_points = int(num_points)
```
Comment (JoelPendleton) [inline stanza/routines/builtins/health_check.py:130]: Same as above

File: stanza/routines/builtins/health_check.py [hunk 10]
```diff
@@ -0,0 +1,657 @@
...
+
+    control_gates = ctx.resources.device.control_gates
+    control_gate_configs = {
+        gate: ctx.resources.device.channel_configs[gate] for gate in control_gates
+    }
+
+    max_voltage_bound = min(
```
Comment (JoelPendleton) [inline stanza/routines/builtins/health_check.py:137]: To make this more general, you should allow each gate to have a different max voltage bound

File: stanza/routines/builtins/health_check.py [hunk 11]
```diff
@@ -0,0 +1,657 @@
...
+        control_gate_configs.values(), key=lambda x: x.voltage_range[1]
+    ).voltage_range[1]
+    min_voltage_bound = max(
```
Comment (JoelPendleton) [inline stanza/routines/builtins/health_check.py:140]: To make this more general, you should allow each gate to have a different min voltage bound

File: stanza/routines/builtins/health_check.py [hunk 12]
```diff
@@ -0,0 +1,657 @@
...
+        ctx.resources.device.jump(prev_voltages_dict, wait_for_settling=True)
+        currents_matrix = []
+        for gate in control_gates:
+            current_voltage = prev_voltages_dict[gate] + delta_V
+            ctx.resources.device.jump({gate: current_voltage}, wait_for_settling=True)
+
+            currents_array = ctx.resources.device.measure(control_gates)
+            current_diff = currents_array - initial_currents_array
+
+            if np.isclose(current_voltage - prev_voltages_dict[gate], 0, atol=1e-10):
```
Comment (JoelPendleton) [inline stanza/routines/builtins/health_check.py:582]: Shouldn't this be done before you set the voltage, not after. It wastes time otherwise

File: stanza/routines/builtins/health_check.py [hunk 13]
```diff
@@ -0,0 +1,657 @@
...
+            delta_V,
+            session,
+        ):
+            # Reset to initial voltages before returning
+            ctx.resources.device.jump(
+                dict(zip(control_gates, initial_voltages, strict=False)),
+                wait_for_settling=True,
+            )
+            return delta_V, True
+
```
Comment (JoelPendleton) [inline stanza/routines/builtins/health_check.py:630]: What happens if a current never exceeds the threshold for a gate? It's possible the gate is very good (does not leak) or it is broken.

Comment (JoelPendleton) [reply stanza/routines/builtins/health_check.py:630]: We should keep track of gates which never leak over both the min and max bounds and show a warning. In the future, we could have another routine which does further testing to determine if this is a broken gate

File: stanza/routines/builtins/health_check.py [hunk 14]
```diff
@@ -0,0 +1,657 @@
...
+    bias_gate: str,
+    bias_voltage: float,
+    session: LoggerSession | None = None,
+) -> dict[str, float]:
+    """Determine the global turn-on voltage by sweeping all control gates simultaneously.
+
+    This BATIS routine sweeps all control gates together from minimum to maximum voltage
+    while measuring current at a specified electrode. The sweep data is analyzed using
+    a pinch-off heuristic to identify the voltage at which the device transitions from
+    depletion to accumulation (turn-on). After finding this voltage, all gates are set
```
Comment (JoelPendleton) [inline stanza/routines/builtins/health_check.py:223]:
```suggestion
    depleted to accumulated (turn-on). After finding this voltage, all gates are set
```

File: stanza/routines/builtins/health_check.py [hunk 15]
```diff
@@ -0,0 +1,657 @@
...
+    step_size: float,
+    bias_gate: str,
+    bias_voltage: float,
+    session: LoggerSession | None = None,
+) -> dict[str, dict[str, Any]]:
+    """Characterize individual reservoir gates by sweeping each while holding others in accumulation.
+
+    This BATIS routine determines the turn-on voltage (pinch-off voltage) for each reservoir gate
+    individually. For each reservoir under test, all other reservoirs are set to 120% of the global
+    turn-on voltage (to ensure they're fully conducting), while the target reservoir is swept from
```
Comment (JoelPendleton) [inline stanza/routines/builtins/health_check.py]:
```suggestion
    turn-on voltage (to ensure they're accumulated), while the target reservoir is swept from
```

File: stanza/routines/builtins/health_check.py [hunk 16]
```diff
@@ -0,0 +1,657 @@
...
+    minimum to maximum voltage. This isolates the behavior of each reservoir and identifies its
+    individual turn-on characteristics.
```
Comment (JoelPendleton) [inline stanza/routines/builtins/health_check.py]:
```suggestion
    individual pinch-off characteristics.
```

File: stanza/routines/builtins/health_check.py [hunk 17]
```diff
@@ -0,0 +1,657 @@
...
+    bias_voltage: float,
+    session: LoggerSession | None = None,
+) -> dict[str, dict[str, Any]]:
+    """Characterize individual finger gates by sweeping each while holding others in accumulation.
+
+    This BATIS routine determines the turn-on voltage (pinch-off voltage) for each finger gate
+    individually. For each finger gate under test, all other finger gates are set to 120% of the
+    global turn-on voltage (to ensure they're fully conducting), while the target gate is swept
+    from minimum to maximum voltage. This isolates the behavior of each finger gate and identifies
+    its individual turn-on characteristics.
```
Comment (JoelPendleton) [inline stanza/routines/builtins/health_check.py]:
```suggestion
    its individual pinch-off characteristics.
```

File: stanza/routines/builtins/health_check.py [hunk 18]
```diff
@@ -0,0 +1,657 @@
...
+    step_size: float,
+    bias_gate: str,
+    bias_voltage: float,
+    session: LoggerSession | None = None,
+) -> dict[str, dict[str, Any]]:
+    """Characterize individual finger gates by sweeping each while holding others in accumulation.
+
+    This BATIS routine determines the turn-on voltage (pinch-off voltage) for each finger gate
+    individually. For each finger gate under test, all other finger gates are set to 120% of the
+    global turn-on voltage (to ensure they're fully conducting), while the target gate is swept
```
Comment (JoelPendleton) [inline stanza/routines/builtins/health_check.py]:
```suggestion
    global turn-on voltage (to ensure they're fully accumulated), while the target gate is swept
```

File: stanza/routines/builtins/health_check.py [hunk 19]
```diff
@@ -0,0 +1,657 @@
...
+        leakage_threshold_count: Maximum number of leaky gate pairs allowed.
+        control_gates: List of control gate names for reporting.
+        delta_V: Current voltage offset being tested (V).
+        session: Optional logger session for recording leakage analysis.
+
+    Returns:
+        True if leakage threshold is exceeded, False otherwise.
+    """
+    is_leaky = leakage_matrix < leakage_threshold_resistance
+    upper_triangle = np.triu(is_leaky, k=1)
```
Comment (JoelPendleton) [inline stanza/routines/builtins/health_check.py:511]: Why upper triangle and not np.where?

File: tests/test_health_check.py [hunk 20]
```diff
@@ -0,0 +1,453 @@
...
+        assert "min_safe_voltage_bound" in result
+        assert isinstance(result["max_safe_voltage_bound"], (int, float))
+        assert isinstance(result["min_safe_voltage_bound"], (int, float))
+
+    def test_leakage_uses_current_std_from_results(self, routine_context):
+        routine_context.results.store("current_std", 5e-11)
+
+        result = leakage_test(
+            routine_context,
+            leakage_threshold_resistance=1e6,
```
Comment (JoelPendleton) [inline tests/test_health_check.py]: Should be 50MOhm threshold

File: tests/test_health_check.py [hunk 21]
```diff
@@ -0,0 +1,453 @@
...
+            leakage_threshold_count=0,
+            control_gates=control_gates,
+            delta_V=0.1,
+            session=session,
+        )
+
+        assert leaked
+        assert len(session.analyses) == 1
+        assert session.analyses[0][0] == "leaky_gate_pairs"
```
Comment (JoelPendleton) [inline tests/test_health_check.py:457]: Double check all code that you test that does curve fitting, actually tests fitting curves of both positive / success cases, e.g. a sigmoid, exponential or straight line, and failure cases such as random noise.

This will be the biggest form of failure across the routines.

**Review by JoelPendleton — CHANGES_REQUESTED** [review]

File: stanza/drivers/qdac2.py (hunk 2 above)
Comment (rjowcq) [reply stanza/drivers/qdac2.py]: I had `prepare_measurement` as a required overwritten function of the `BaseInstrument` class, but after thinking about the implications, we should just make it optional. I've removed it from the qdac2 driver.

File: stanza/drivers/qdac2.py (hunk 4 above)
Comment (rjowcq) [reply stanza/drivers/qdac2.py]: added

File: stanza/analysis/fitting.py (hunk 3 above)
Comment (rjowcq) [reply stanza/analysis/fitting.py:17]: yes, these bounds assume normalized data.  There are already tests in @test_analysis_fitting.py - I've added more to test a wide array of edge cases (including poor fits based on the RMSE criterion).

File: tests/test_health_check.py (hunk 21 above)
Comment (rjowcq) [reply tests/test_health_check.py:457]: addressed (see comment above).

File: stanza/analysis/fitting.py [hunk 22]
```diff
@@ -0,0 +1,231 @@
...
+        np.ndarray: Pinchoff current f(x) output
+    """
+    return a * (1 + np.tanh(b * x + c))
+
+
+def derivative_extrema_indices(x: np.ndarray, y: np.ndarray) -> tuple[int, int, int]:
+    """
+    Return the indices of key voltages for pinchoff curves.
+
+    vp corresponds to the pinchoff state (low current)
```
Comment (JoelPendleton) [inline stanza/analysis/fitting.py]: This is out of date

File: stanza/routines/builtins/health_check.py (hunk 19 above)
Comment (rjowcq) [reply stanza/routines/builtins/health_check.py:511]: `np.triu` is generally much faster than np.where, since it doesn't need to compute each index and extract.

File: stanza/routines/builtins/health_check.py (hunk 6 above)
Comment (rjowcq) [reply stanza/routines/builtins/health_check.py:28]: done

File: stanza/routines/builtins/health_check.py (hunk 7 above)
Comment (rjowcq) [reply stanza/routines/builtins/health_check.py:128]: yeah - due to how yaml parses specific values (they may end up as floating points.  For example `50e6` ends up as a floating point, since it is in scientific notation).  This is just a sanity check.

File: stanza/routines/builtins/health_check.py (hunk 13 above)
Comment (rjowcq) [reply stanza/routines/builtins/health_check.py:630]: we keep track of all gates, and log the leaky gates in case of a failure, so that should give complete information for all leaky gate failures.

File: stanza/routines/builtins/health_check.py (hunk 10 above)
Comment (rjowcq) [reply stanza/routines/builtins/health_check.py:137]: this can be done in the future - Since we'll need to keep track every single gate bound individually, this will make the leakage test much more complex then it already is.  Let's assume that all gates have the same gate bounds for now.

File: stanza/routines/builtins/health_check.py (hunk 11 above)
Comment (rjowcq) [reply stanza/routines/builtins/health_check.py:140]: same as above.

**Review by JoelPendleton — CHANGES_REQUESTED** [review]

File: stanza/analysis/fitting.py
Comment (brandonseverin) [inline stanza/analysis/fitting.py:1]: Add more verbose docstrings across the board

File: stanza/analysis/preprocessing.py [hunk 23]
```diff
@@ -0,0 +1,7 @@
+import numpy as np
+
+
+def normalize(a: np.ndarray) -> np.ndarray:
```
Comment (brandonseverin) [inline stanza/analysis/preprocessing.py:4]: make this more verbose docstring

File: stanza/base/channels.py
Comment (brandonseverin) [inline stanza/base/channels.py:1]: remove GPIO references for this merge - put on separate branch

File: stanza/configs/devices/device.sample.yaml
Comment (brandonseverin) [inline stanza/configs/devices/device.sample.yaml:1]: remove gpio reference put on separate branch - place on separate branch
- replace MUX with GPIO to be more general.

File: stanza/device.py
Comment (brandonseverin) [inline stanza/device.py:1]: More verbose doscstrings

File: stanza/models.py
Comment (brandonseverin) [inline stanza/models.py:1]: more verbose docstrings

File: tests/test_health_check.py
Comment (brandonseverin) [inline tests/test_health_check.py:1]: Please add atleast a oneline docstring for each test and class

File: tests/test_qdac2.py
Comment (brandonseverin) [inline tests/test_qdac2.py:1]: missing docstrings

**Review by brandonseverin — CHANGES_REQUESTED** [review]
be more verbose with docstrings - every module, function and class should have one. overall good -

File: stanza/analysis/preprocessing.py (hunk 23 above)
Comment (rjowcq) [reply stanza/analysis/preprocessing.py:4]: done

File: stanza/device.py
Comment (rjowcq) [reply stanza/device.py:1]: done

**Review by brandonseverin — APPROVED** [review]

**Review by JoelPendleton — APPROVED** [review]
